### ISSUE TITLE
unable to get parsed config in code

### ISSUE THREAD

**Issue by Morriz** [strongloop/loopback#2062]
How do we access the parsed config in a straightforward manner in the code? I can't find any documentation on this. Is it something like:

```
loopback.getConfig()
```
?

**Comment by Morriz** [comment]
I know I can require the app from the server.js, which should give access to it, but I wish it would be booted and ready to consume config settings from. The config should be resolved synchronously and made available before the server boots. That way we can bootstrap modules in userland with the config without having to wait for that 'hold your horses' approach that makes us all wait for everything to be resolved. It has way to many pains. Seperating out and publishing stuff in the right order would alleviate much of those pains.

**Comment by gunjpan** [comment]
@Morriz : You can access the configs in the `boot` scripts under `/server/boot` such as:
```javascript:boot-script.js
module.exports = function (app) {
  app.get('loopback-component-explorer'); // returns value from `component-config.json`
  app.get('restApiRoot'); // returns value of `restApiRoot` property from `config.json`
};
```
[Boot scripts](https://docs.strongloop.com/display/public/LB/Defining+boot+scripts) is the place where you would like to customize boot strapping as these scripts are executed before server starts and `app` object has all `configs`.
_(also [`loopback-boot`](https://apidocs.strongloop.com/loopback-boot/#boot) is something that would provide more information how boot works)_

Is that something you're looking for?! Hope that helps, thank you!

**Comment by Morriz** [comment]
I wish to be able to reach the same config from other locations, and think loopback should accomodate as such. Because I feel it's dirty to have every user of loopback create a script that runs during boot, to save the config to global or shared lib, which is what you suggest in so many words.

And, is that even possible? Can I get the WHOLE config with app.get()? Otherwise your suggestion would be pointless.

**Comment by Morriz** [comment]
Plus, WHY would loopback not have at least the synchronous part of the config ready immediately? I feel its a bad decision to treat it all as async, just because there might be some async config reads? Please be smart about such things, and less crude to the users.

Let the users know when they request async stuff they have to wait for it. Like with any code. Don't force them to wait for very late calls to read configs that can be read sync.

**Comment by Morriz** [comment]
I believe a rewrite of the config parsing is in order, and at least one page of documentation would be nice.

**Comment by Morriz** [comment]
I got my config this way:

```
const config = ConfigLoader.loadAppConfig(path.join(__dirname, '../../server'), process.env.NODE_ENV)
```

Why not make that available pre-boot? Or at least document that...

**Comment by bajtos** [comment]
1) We store data from `config.json` via express' `app.set`. You can retrieve all settings via `app.settings`.

2) The main problem of `loopback.getConfig()` is that it would be a global singleton, which does not work when there is multiple loopback apps in the same process (or what's even worse, when there are multiple copies of `loopback` in node_modules tree. We have been bitten by this several times, in fact we are working on moving model & datasource registration out of `loopback.*` to `app.*`.

> WHY would loopback not have at least the synchronous part of the config ready immediately

Consider the following example:

```js
var loopback = require('loopback');  
var booster = require('./lib/booster');
var app = loopback();
app.setConfig({...}); // pretend this is called by loopback-boot
booster(app);
```

If you call `loopback.getConfig()` from `lib/booster`, then you would get a different data than when you call `loopback.getConfig()` from `booster`.

This also raises a question related to 2) - how is your code calling `loopback.getConfig` going to get the value of `loopback`? `require('loopback')` works only if you call it in the top-level project.

> I got my config this way:
> (...)
> Why not make that available pre-boot? Or at least document that...

It is too coupled to the file layout in the project and it does not work in the client/browser.

@Morriz could you please provide more details on what exactly are you trying to accomplish? We may be able to point you to a solution that would satisfy your needs.

For example, you can access app-wide settings in models via `MyModel.app.get()` or `MyModel.app.settings` (but only after the model was attached to an application via `app.model`, which is typically called by loopback-boot).

As for `loopback.getConfig`, I am personally opposed to that idea, however you can easily implemet it yourself - just add the following line to your `server/server.js`:

```js
//...
var app = module.exports = loopback();
loopback.getConfig = function() { return app.settings; } // <-- add this line
// ...
```

**Comment by Morriz** [comment]
First of all tnx for such a detailed response. I did not look as far as the scenario you described.

But, would it not be possible to treat 'multiple loopback apps' as an edge case, and create documentation as such? And provide sane defaults?

I think most people don't have multiple apps running (especially now that we are in the micro service age).

All I wanted was to know the config directly before boot, and got that with the solution I gave:
```ConfigLoader.loadAppConfig(path.join(__dirname, '../../server'), process.env.NODE_ENV)```
(change paths according to your setup of course)

I think some convenience method could be exposed that does just that: read the config synchronously for an app, like such:
```loopback.getServerConfig(pathWhereServerDotJsLives)```
(and memoize it)
